docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,63 @@
+# ADR 016: Validator Consensus Key Rotation
+
+## Changelog
+
+- 23-10-2019: initial draft
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction(RotateValConsensusKey) that the new consensus key is now coupled with the validator operator with signature from validator wallet key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key for signing votes.
```

Comment:
We should also note that the KMS/HSM should act accordingly (with the correct key) at a specific H, where H is the height after the tx is committed. Maybe this will be tricky?

Reply:
do you mean that the state machine combined with KMS/HSM should support auto-consensus-key-switch feature? I think it is not only tricky, but also too risky to be automated. I think key rotation deserves some manual operations considering related double-signing risk.

Reply:
No no, I mean operators should take note that their KMS/HSM act and operate accordingly. Meaning, when you submit `MsgRotateConsPubKey` and it gets confirmed at height `H`, that their setup should start using the new key at height `H+1` and not the old key.

Reply:
Okay. I will make it clearer.

Reply:
done!